src/ethereum/state_oracle/interface.py

```diff
@@ -0,0 +1,213 @@
+
+
+class MerkleOracle(Protocol):
+    """
+    Oracle interface for Merkle Patricia Trie based state.
+    """
+
+    def get_account(
+        self, address: Address  # noqa: U100
+    ) -> Account:  # noqa: U100
+        """
+        Get account information for the given address.
+
+        `address` is the account address to retrieve.
+
+        Returns EMPTY_ACCOUNT if the account doesn't exist.
+        """
+
+    def get_account_optional(
+        self, address: Address  # noqa: U100
+    ) -> Optional[Account]:  # noqa: U100
+        """
+        Get account information for the given address.
+
+        `address` is the account address to retrieve.
+
+        Returns None if the account doesn't exist.
+        Use this when you need to distinguish between non-existent and
+        empty accounts.
+        """
+
+    def get_storage(
+        self, address: Address, key: Bytes32  # noqa: U100
+    ) -> Bytes32:  # noqa: U100
+        """Get storage value at `key` for the given `address`."""
+
+    def state_root(self) -> Bytes32:  # noqa: U100
+        """Compute and return the current state root."""
+
+    def get_storage_original(
+        self, address: Address, key: Bytes32  # noqa: U100
+    ) -> Bytes32:  # noqa: U100
+        """
+        Get original storage value before current transaction started.
+
+        `address` is the contract address and `key` is the storage slot.
+
+        This is required for SSTORE gas calculations per EIP-2200.
+        The implementation should use state snapshots/checkpoints to
+        track pre-transaction values.
+        TODO: The oracle does not have a `begin_transaction` method,
+        TODO: so it kind of breaks here.
+
+        Parameters
+        ----------
+        address : Bytes20
+            Contract address
+        key : Bytes32
+            Storage slot key
+
+        Returns
+        -------
+        Bytes32
+            Original storage value as 32-byte value
+        """
+
+    def set_storage_value(
+        self, address: Address, key: Bytes32, value: Bytes32  # noqa: U100
+    ) -> None:
+        """
+        Set individual storage value.
+
+        `address` is the contract address, `key` is the storage slot,
+        and `value` is the new storage value.
+
+        Parameters
+        ----------
+        address : Bytes20
+            Contract address
+        key : Bytes32
+            Storage slot key
+        value : Bytes32
+            Storage value as 32-byte value
+        """
+
+    def account_has_code_or_nonce(
+        self, address: Address  # noqa: U100
+    ) -> bool:  # noqa: U100
+        """
+        Check if account has non-zero code or nonce.
+
+        Used during contract creation to check if address is available.
+        """
+
+    def account_has_storage(self, address: Address) -> bool:  # noqa: U100
+        """
+        Check if account has any storage slots.
+
+        Used during contract creation to check if address is available.
+        """
```

Comment:
As noted in the PR description; methods like these can be moved into something like an "EVMOracle", since they can be implemented using get_account plus a few extra LOC to check if there is code/nonce/storage.

The user would then only implement MerkleOracle, which then gets wrapped in an EVMOracle so that the EVM can use the relevant methods

Reply:
This is the motivation behind the minimal state interface in #1209. Everything that isn't off the shelf that can be done in the specs should be done there, even if the performance is terrible.